### DISCUSSION TITLE
Proper usage of cfg_attr

### DISCUSSION THREAD

**Discussion by jafioti** [diesel-rs/diesel#2942] [Q&A] [answered]
I'm trying to make diesel integration a feature in my crate, and it almost works with a feature flag, where I can hide all diesel #[derive] statements behind a cfg_attr, except for the #[primary_key()]. Here is one of my structs:

```
#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(feature = "database", derive(Identifiable, Insertable, Queryable, Associations))]
#[cfg_attr(feature = "database", table_name="blog_posts")]
#[cfg_attr(feature = "database", primary_key(title))]
pub struct BlogPost {
    pub title: String,
    pub picture_url: String,
    pub date: i32,
    pub body: String,
}
```

This doesn't work because when "database" feature is activated, diesel doesn't see the primary_key line. Am I doing this correctly? Is this possible?

**Comment by weiznich** [comment, marked as answer]
Can you provide the following information, as it is otherwise really hard to debug this problem:
* The exact error message that occurs
* The version of diesel, rustc and cargo you are using
* A minimal complete example, so a complete rust file + a corresponding minimal toml file

**Reply by jafioti** [reply]
Cargo.toml:
```
[package]
name = "database"
version = "1.0.1"
edition = "2018"

[features]
default = ["database"]
database =  ["diesel"]

[dependencies]
diesel = { version = "1.4.4", features = ["sqlite"], optional = true }
```

main.rs:
```
#[cfg(feature = "database")]
use diesel::Queryable;
#[cfg(feature = "database")]
use crate::schema::blog_posts;

#[cfg_attr(feature = "database", derive(Identifiable, Insertable, Queryable, Associations))]
#[cfg_attr(feature = "database", table_name="blog_posts")]
#[cfg_attr(feature = "database", primary_key(title))]
pub struct BlogPost {
    pub title: String,
    pub picture_url: String,
    pub date: i32,
    pub body: String,
}
```

**Reply by weiznich** [reply]
I've tried your example, but I failed to reproduce the described error. Please share the error message and a complete setup to reproduce the problem.

**Reply by jafioti** [reply]
Actually after I tried that example, it seems that only rust-analyzer throws an error, but rustc does not when built. Are you seeing this effect?

**Reply by weiznich** [reply]
I did not check that, but if that's the case this is clearly a rust-analyzer bug then. I would suggest you report the issue in their issue tracker. Feel free to cc me from there.
